### ISSUE TITLE
Add create to Storage that takes an InputStream

### ISSUE THREAD

**Issue by mziccard** [GoogleCloudPlatform/gcloud-java#172]
`Storage` class now has a method to create an object:
```Java
create(BlobInfo blobInfo, final byte[] content, BlobTargetOption... options)
```
I know that we already provide a `BlobWriteChannel` but it might be useful to also have:
```Java
create(BlobInfo blobInfo, InputStream content, BlobTargetOption... options)
```
for uploading small files without reading their bytes. Thoughts?

**Comment by aozarov** [comment]
I don't like having various flavors of the same method/functionality, especially when the difference is so subtle because at some point it makes the API less readable (what about adding also `ByteBuffer` option,....). This method should be used when content can live in memory.  I picked `bytes[]` because I thought it is a good common denominator but could be convinced to pick something else.

It is fairly simple to convert a `InputStream` to `byte[]` (e.g. by using [Guava](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/io/ByteStreams.html#toByteArray(java.io.InputStream))).

**Comment by mziccard** [comment]
I see your point but let me play the devil's advocate here.
It could argued that having both methods makes the library easier to use, as for instance, a user with a small `FileInputStream` is not required to read the stream to a byte array (even if it is farly simple) but can upload that with a single library call.

**Comment by aozarov** [comment]
I assume that your suggestion includes also using a different RPC (apiary with an inputstream for media uploader - see [here](https://github.com/google/google-api-java-client/blob/master/google-api-client/src/main/java/com/google/api/client/googleapis/media/MediaHttpUploader.java)) rather than us reading the content into a byte array and using the same rpc?

It would make more sense in that case even though I think generally we should recommend using resumable uploads for big files (via  BlobWriteChannel) and non-resumables for small ones. Can you check if current implementation of the MediaUploader will really chunk when using direct/non-resumable upload rather than itself reading all in memory before sending the request?

**Comment by mziccard** [comment]
Yes I was planning on adding a different RPC which makes uses of `InputStreamContent` (see [here](https://developers.google.com/api-client-library/java/google-http-java-client/reference/1.20.0/com/google/api/client/http/InputStreamContent)) rather than `ByteArrayContent`.  

From the docs of `MediaHttpUploader` I read:
>  For resumable uploads, when the media content length is known, if the provided InputStream has InputStream.markSupported() as false then it is wrapped in an BufferedInputStream to support the InputStream.mark(int) and InputStream.reset() methods required for handling server errors. If the media content length is unknown then each chunk is stored temporarily in memory. This is required to determine when the last chunk is reached.

Which makes me think that is always tries to use resumable uploads (even for `ByteArrayContent`).

**Comment by aozarov** [comment]
Yes, unless we pass `true` to `setDirectUploadEnabled`. see [here](https://developers.google.com/api-client-library/java/google-api-java-client/reference/1.20.0/com/google/api/client/googleapis/media/MediaHttpUploader#setDirectUploadEnabled(boolean)). So my understanding (but I didn't dig deep) is that using this apiary client resumable downloads  can work in chunks (for all input types) but for non-resumable (which is what I think we want here) one need to specify direct-upload and I was not sure if that will not load the complete content to memory before sending it (conceptually it does not have to). Can you verify these assumptions?

**Comment by mziccard** [comment]
When `directUploadEnabled` is set to true, whether data are read in chunks or not depends on the type of `XXXContent`. In the case of all subclasses of `AbstractInputStreamContent` data reads happen in chunks of 2048 bytes.

**Comment by aozarov** [comment]
You mean this [`AbstractInputStreamContent`](https://github.com/google/google-http-java-client/blob/dev/google-http-client/src/main/java/com/google/api/client/http/AbstractInputStreamContent.java#L72)? As this
one seems to write all in one write... (or do they override something?)

**Comment by mziccard** [comment]
I might be wrong but doesn't `AbstractInputStreamContent` use `IOUtils.copy` which then calls [`BufferStreams.copy`](https://github.com/google/google-http-java-client/blob/346a896b201a784533fab6f5d0214dc62db21a70/google-http-client/src/main/java/com/google/api/client/util/ByteStreams.java#L45)? That should work in chunks (4096B actually).

**Comment by aozarov** [comment]
Ok, if you are sure that the destination (`out`) is streamed when `directUploadEnable` is true (which would make sense) rather than buffered and sent as one blob then I think your proposal of adding this method (and an extra RPC) make sense. Maybe we can use one RPC with InputStream then....

**Comment by aozarov** [comment]
Implemented by #186
